### PR TITLE
Reverse order of arguments to make HttpClient instrumentation enablement less error prone

### PR DESCRIPTION
**Pull request by cijothomas**
Addresses https://github.com/open-telemetry/opentelemetry-dotnet/issues/1787#issuecomment-773547935

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs [hunk 1]
```diff
@@ -30,13 +30,13 @@ public static class TracerProviderBuilderExtensions
...
         /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
-        /// <param name="configureHttpClientInstrumentationOptions">HttpClient configuration options.</param>
         /// <param name="configureHttpWebRequestInstrumentationOptions">HttpWebRequest configuration options.</param>
+        /// <param name="configureHttpClientInstrumentationOptions">HttpClient configuration options.</param>
         /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
         public static TracerProviderBuilder AddHttpClientInstrumentation(
             this TracerProviderBuilder builder,
-            Action<HttpClientInstrumentationOptions> configureHttpClientInstrumentationOptions = null,
-            Action<HttpWebRequestInstrumentationOptions> configureHttpWebRequestInstrumentationOptions = null)
+            Action<HttpWebRequestInstrumentationOptions> configureHttpWebRequestInstrumentationOptions = null,
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:38]: Doesn't seem any less confusing or error-prone to me. IMO the solution is to not use "object" in the enrich callback signatures. Type safety/compile-time checking is our friend!

**Comment by utpilla** [comment]
What's the most common usage for adding `HttpClientInstrumentation` on .NET Framework? Do they usually only add `HttpClientInstrumentationOptions` and not `HttpWebRequestInstrumentationOptions` ? If yes, then this might make it less error-prone as most users only provide `HttpClientInstrumentationOptions` .

File: src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:38]: @CodeBlanch I think the issue here is that many users call `AddHttpClientInstrumentation(options => {...})` on .NET Framework which is how it's called on .NET Core and the README also only has examples with this usage.

The problem is that this also compiles for .NET Framework which has different signature for `AddHttpClientInstrumentation` that take two arguments `HttpClientInstrumentationOptions` and `HttpWebRequestInstrumentationOptions` which are null by default. So when they setup `HttpWebRequestInstrumentationOptions` and call `AddHttpClientInstrumentation(options => {...})` it won't work as they should be calling `AddHttpClientInstrumentation(null, options => {...})`

**Comment by CodeBlanch** [comment]
Here's some code to demo my point:

```csharp
public Action<Activity, string, object> Enrich { get; set; }

options.Enrich = (Activity activity, string eventName, object myLazyObject) => {
   (HttpWebRequest)myLazyObject; // Exception thrown if myLazyObject is HttpRequestMessage
}
```

vs

```csharp
public Action<Activity, string, HttpRequestMessage> Enrich { get; set; }

options.Enrich = (Activity activity, string eventName, HttpWebRequest myLazyObject) => {
    // Won't compile because HttpWebRequest in delegate does not match enrich signature.
    // User will have to fix the params order to target the correct options.
}
```

IMO the issue is really due to...

```csharp
public class HttpClientInstrumentationOptions
{
  public Action<Activity, string, object> Enrich { get; set; }
}

public class HttpWebRequestInstrumentationOptions
{
   public Action<Activity, string, object> Enrich { get; set; }
}
```

If that object was _something_ typed, happy days.

**Comment by cijothomas** [comment]
@CodeBlanch Sorry I wasn't sharing more details in the PR description. While I agree about the Enrich part, this PR was an attempt to make the AddHttpClientInstrumentation() less error prone for a different scenario! (not related to Enrich at all)

Eg: If a user wants to SetHttpFlavor=true, then the following code in .NET Framework doesn't achieve it.

```
AddHttpClientInstrumentation(options =>
                 {
                     options.SetHttpFlavor = true;
                 })
```

Instead the following code is required to achieve what's needed.
```
AddHttpClientInstrumentation(null, options =>
                 {
                     options.SetHttpFlavor = true;
                 })
```

This PR addresses this part only. I guess the bigger question which I was trying to find answer was - do we need both `HttpClientInstrumentationOptions` and `HttpWebRequestInstrumentationOptions` for .NET FW? I assume the answer is yes, but the common scenario in .NET FW is to use HttpWebRequest and hence making the 1st param `HttpWebRequestInstrumentationOptions` makes it less likely to be error-prone.

**Comment by CodeBlanch** [comment]
> I guess the bigger question which I was trying to find answer was - do we need both HttpClientInstrumentationOptions and HttpWebRequestInstrumentationOptions for .NET FW?

I was going to bring that up but I didn't want to open a can of worms 🥫 If we only had the HttpWebRequest options for .NET Framework, I'm pretty sure it would still be called when using HttpClient! Because HttpClient on .NET Framework just wraps HttpWebRequest. The only weird thing would be users working with HttpRequestMessage/HttpResponseMessage via HttpClient would actually be receiving the underlying HttpWebRequest/HttpWebResponses objects in the enrich callback. That could be surprising! But might be the better option?

I can verify this if you would like, just LMK.

**Comment by cijothomas** [comment]
> > I guess the bigger question which I was trying to find answer was - do we need both HttpClientInstrumentationOptions and HttpWebRequestInstrumentationOptions for .NET FW?
> 
> I was going to bring that up but I didn't want to open a can of worms 🥫 If we only had the HttpWebRequest options for .NET Framework, I'm pretty sure it would still be called when using HttpClient! Because HttpClient on .NET Framework just wraps HttpWebRequest. The only weird thing would be users working with HttpRequestMessage/HttpResponseMessage via HttpClient would actually be receiving the underlying HttpWebRequest/HttpWebResponses objects in the enrich callback. That could be surprising! But might be the better option?
> 
> I can verify this if you would like, just LMK.

I think this PR - https://github.com/open-telemetry/opentelemetry-dotnet/pull/1982, is better.
If a user is on .NET FW or .NET Core, they always have 
AddHttpClientInstrumentation()

on .NET FW, options are HttpWebRequestInstrumentationOptions
on .NET Core, options are HttpClientInstrumentationOptions

The doc (https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.Http#filter and https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.Http#enrich)  should make it clear which type to expect for .NET FW vs .NET Core. 

Let me know your thoughts. If no objections, I will abandon this PR and take #1982 

The lack of strong type in Enrich - will come back to this separately. We can modify it similar to Filtet (which is strong typed).

**Comment by cijothomas** [comment]
Closing in favor of #1982

**Pull request closed without merging**